### PR TITLE
Updates developer documentation

### PR DESCRIPTION
**Pull request by fvclaus**
I hope this improves the development setup for newcomers. A few things things that gave me trouble:
- The documentation about the variables seems outdated. There is no variable to specify the target platform anymore.
- Using Oomph to install Eclipse and setup Buildship, did not work. It started Eclipse, but did not import Buildship. After rereading the documentation just now, I learned about Help > Perform setup task, which displays the following error:
import_error
The `buildship/` folder only contained one `tooling.target` file. After manually checking out the source code, Perform setup task finished and the project was imported properly. Therefore, I also added how to import the project with an existing instance.
 - I had no JDK setup on my systems (its Windows, I do my development in a WSL). Eclipse downloaded JDK 17 and prefilled the JRE 1.8 and JRE 11 with the JDK 17 path. It's a very long path and I trusted Eclipse to do the right thing. This lead to one of the craziest exception I have ever seen. `Class.newInstance` threw a `ClassDefNotFoundError` with the Class it was trying to instantiate. In the debugger, I could see the Class correctly loaded and could interact with it, but calling `.newInstance` would lead to a `ClassDefNotFoundError`. After a few hours, I discovered that my Eclipse used JDK 17 to run the tests (JRE11 Execution Environment used my JDK 17). I added a small paragraph to highlight the importance of these settings.
 - On my system, Eclipse uses CP1252 as text file encoding, this causes Problem in the one file that uses a unicode character, a soft-hyphen in `PatternUtilsTest`. The full message is:
 ```
Groovy:Unexpected character 0xAD (SOFT HYPHEN) at column 21
```

### CODE REVIEW

**Comment by FieteO** [comment]
I am such a newcomer. I have opened a ticket (#1149) that describes my setup troubles.
I will try it again following your changes when I have time.

**Comment by fvclaus** [comment]
@FieteO, good to hear that this is already useful to somebody. I assume by changes, you refer to my changes in #1154? I wouldn't recommend checking out my branch. I wasn't able to add another sub project (testdependencies) to the buildship project created from the Oomph catalog. And I am not sure the branch works for projects that were checked out using the Oomph catalog. Somebody with more knowledge about Oomph has to take a look at that. What I ended up doing is to use the (deprecated) manual setup that is described in `development/Setup.md`.

Instead, I would recommend to make some "minor" changes to be able to run the tests. Here are all the steps are necessary steps:

### Change JRE to 11
setup-3
Make sure the JRE referenced here is actually a JRE 11. This can be verified by clicking on "Environments"
setup-2
My JRE 11 actually directs to a JDK 17.
setup-21
If you run the tests with JDK 17, you will get the following exception:
```
java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.vmplugin.v7.Java7
	at org.codehaus.groovy.vmplugin.VMPluginFactory.<clinit>(VMPluginFactory.java:43)
	at org.codehaus.groovy.reflection.GroovyClassValueFactory.<clinit>(GroovyClassValueFactory.java:35)
	at org.codehaus.groovy.reflection.ClassInfo.<clinit>(ClassInfo.java:109)
	at org.codehaus.groovy.reflection.ReflectionCache.getCachedClass(ReflectionCache.java:95)
	at org.codehaus.groovy.reflection.ReflectionCache.<clinit>(ReflectionCache.java:39)
	at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.registerMethods(MetaClassRegistryImpl.java:209)
	at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.<init>(MetaClassRegistryImpl.java:107)
	at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.<init>(MetaClassRegistryImpl.java:85)
	at groovy.lang.GroovySystem.<clinit>(GroovySystem.java:36)
	at org.codehaus.groovy.runtime.InvokerHelper.<clinit>(InvokerHelper.java:86)
	at groovy.lang.GroovyObjectSupport.getDefaultMetaClass(GroovyObjectSupport.java:59)
	at groovy.lang.GroovyObjectSupport.<init>(GroovyObjectSupport.java:32)
	at groovy.lang.Closure.<init>(Closure.java:217)
	at groovy.lang.Closure.<init>(Closure.java:234)
	at groovy.lang.Closure$1.<init>(Closure.java:199)
	at groovy.lang.Closure.<clinit>(Closure.java:199)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at org.spockframework.runtime.extension.builtin.IgnoreIfExtension.createCondition(IgnoreIfExtension.java:52)
	at org.spockframework.runtime.extension.builtin.IgnoreIfExtension.doVisit(IgnoreIfExtension.java:43)
	at org.spockframework.runtime.extension.builtin.IgnoreIfExtension.visitFeatureAnnotation(IgnoreIfExtension.java:39)
	at org.spockframework.runtime.extension.builtin.IgnoreIfExtension.visitFeatureAnnotation(IgnoreIfExtension.java:31)
	at org.spockframework.runtime.ExtensionRunner.doRunAnnotationDrivenExtensions(ExtensionRunner.java:95)
	at org.spockframework.runtime.ExtensionRunner.runAnnotationDrivenExtensions(ExtensionRunner.java:74)
	at org.spockframework.runtime.ExtensionRunner.runAnnotationDrivenExtensions(ExtensionRunner.java:51)
	at org.spockframework.runtime.ExtensionRunner.run(ExtensionRunner.java:41)
	at org.spockframework.runtime.Sputnik.runExtensionsIfNecessary(Sputnik.java:90)
	at org.spockframework.runtime.Sputnik.getDescription(Sputnik.java:54)
	at org.junit.runners.model.RunnerBuilder.configureRunner(RunnerBuilder.java:81)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:72)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.requests.ClassRequest.createRunner(ClassRequest.java:28)
	at org.junit.internal.requests.MemoizingRequest.getRunner(MemoizingRequest.java:19)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.createUnfilteredTest(JUnit4TestLoader.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.createTest(JUnit4TestLoader.java:76)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.loadTests(JUnit4TestLoader.java:49)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:513)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:74)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at java.base/java.lang.Thread.run(Thread.java:833)
log4j:WARN No appenders could be found for logger (org.apache.http.client.protocol.RequestAddCookies).
log4j:WARN Please initialize the log4j system properly.
Exception in thread "Plug-in Tests Runner" org.spockframework.util.InternalSpockError: Failed to instantiate spec 'InvocationCustomizerTest'
	at org.spockframework.runtime.BaseSpecRunner.createSpecInstance(BaseSpecRunner.java:110)
	at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:62)
	at org.spockframework.runtime.Sputnik.run(Sputnik.java:63)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:74)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.runtime.InvokerHelper
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallStaticSite(CallSiteArray.java:74)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallSite(CallSiteArray.java:161)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at org.eclipse.buildship.core.internal.test.fixtures.ProjectSynchronizationSpecification.<clinit>(ProjectSynchronizationSpecification.groovy:37)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
```
This is the NoClassDefFoundError I referenced above.

### Remove oomph.tests from the loaded plugins
This is what causes the duplicated Spock library exception, because the Spock library exists twice. Once in `core.tests/libs` and once in `oomph.test/libs`
setup-1


### Fix ambiguos method exception
Now the tests run, but many will fail with

```
groovy.lang.GroovyRuntimeException: Ambiguous method overloading for method java.io.File#<init>.
Cannot resolve which method to invoke for [null] due to overlapping prototypes between:
	[class java.net.URI]
	[class java.lang.String]
	at org.eclipse.buildship.core.internal.test.fixtures.WorkspaceSpecification.setup(WorkspaceSpecification.groovy:69)
```
This is because of an JDK API change. You need to replace lines 69 and 70 in `/org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/internal/test/fixtures/WorkspaceSpecification.groovy` with 

```
        EclipseVmUtil.findOrRegisterVM("8", new File((String) System.getProperty("jdk8.location")))
        EclipseVmUtil.findOrRegisterVM("11", new File((String) System.getProperty("jdk11.location")))   
```

### Add jdk8.location and jdk11.location parameters

setup-5

File: docs/development/Setup.md [hunk 1]
```diff
@@ -56,11 +62,13 @@ finish before doing anything! Once finished, you should see the workspace below.
...
+#### Text file encoding
+
+If your are using Windows, make sure the text file encoding is set to `UTF-8`. This can be done under Windows > Preferences > General > Workspace > Text file encoding. Otherwise you might encounter `Unexpected character` errors.
+
 #### Change the target platform
 
-The target platform selected during the installation can be changed, even if it is a bit hidden. In the main menu, click _Help_ > _Perform Setup tasks_ then click _Back_. Select
-the _Show all variables_ checkbox and change the _Buildship Target Platform_ to the desired version. Click _Next_ and _Finish_, then wait for the target platform to download. Once
-finished, the source code will be compiled against the new target platform.
+The target platform selected during the installation can be changed, even if it is a bit hidden. Open one of the `tooling-eXXX.target` files and select `Set as Active Target Platform`
```
Comment (FieteO) [inline docs/development/Setup.md:72]: I think what is yet missing here is the step to set the `jdk8.location` and `jdk11.location` (in `~/.gradle/gradle.properties`) as per @donat's [comment](https://github.com/eclipse/buildship/pull/1154#pullrequestreview-963599095)

Comment (FieteO) [inline docs/development/Setup.md:72]: I think what is yet missing here is the step to set the `jdk8.location` and `jdk11.location` (in `~/.gradle/gradle.properties`) as per @donat's [comment](https://github.com/eclipse/buildship/pull/1154#pullrequestreview-963599095)

**Comment by FieteO** [comment]
I think I am still facing an issue of not using the correct java versions. I have both tried the deprecated manual setup an d the oomph-based one:

## Oomph
I have followed the instructions in the Setup.md and also made the changes you described above

> @FieteO, good to hear that this is already useful to somebody. I assume by changes, you refer to my changes in #1154? I wouldn't recommend checking out my branch. I wasn't able to add another sub project (testdependencies) to the buildship project created from the Oomph catalog. And I am not sure the branch works for projects that were checked out using the Oomph catalog. Somebody with more knowledge about Oomph has to take a look at that. What I ended up doing is to use the (deprecated) manual setup that is described in `development/Setup.md`.
> 
> Instead, I would recommend to make some "minor" changes to be able to run the tests. Here are all the steps are necessary steps:

Executing the tests with **Run As -> JUnit Plug-in Tests** fails with
``` bash
Caused by: java.lang.ClassCastException: class jdk.internal.loader.ClassLoaders$PlatformClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$PlatformClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
	at org.gradle.initialization.DefaultJdkToolsInitializer.initializeJdkTools(DefaultJdkToolsInitializer.java:41)
	at org.gradle.api.internal.tasks.CompileServices$GradleScopeCompileServices.configure(CompileServices.java:55)
	at jdk.internal.reflect.GeneratedMethodAccessor53.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
	at org.gradle.internal.service.ReflectionBasedServiceMethod.invoke(ReflectionBasedServiceMethod.java:34)
	at org.gradle.internal.service.DefaultServiceRegistry.applyConfigureMethod(DefaultServiceRegistry.java:203)
	... 109 more
```

## Manual setup
I have set the `jdk8.location` and `jdk11.location` environment variables both in `~/.gradle/gradle.properties` and in the `gradle.properties` of the project itself:
``` bash
fiete@ubu:~/buildship-master/git/buildship$ cat ~/.gradle/gradle.properties 
# required properties for the Buildship build
jdk8.location=/usr/lib/jvm/java-1.8.0-openjdk-amd64
jdk11.location=/usr/lib/jvm/java-1.11.0-openjdk-amd64
```
Nevertheless, I get the same exception as for the oomph-based setup:
``` bash
fiete@ubu:~/buildship-master/git/buildship$ ./gradlew clean build --stacktrace

FAILURE: Build failed with an exception.

* What went wrong:
Could not create an instance of type org.gradle.invocation.DefaultGradle_Decorated.
> Could not configure services using GradleScopeCompileServices.configure().

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type org.gradle.invocation.DefaultGradle_Decorated.
        at ...
        ... 49 more
Caused by: org.gradle.internal.service.ServiceLookupException: Could not configure services using GradleScopeCompileServices.configure().
        at ...
        ... 49 more
Caused by: java.lang.ClassCastException: class jdk.internal.loader.ClassLoaders$PlatformClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$PlatformClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
        at org.gradle.initialization.DefaultJdkToolsInitializer.initializeJdkTools(DefaultJdkToolsInitializer.java:40)
        at org.gradle.api.internal.tasks.CompileServices$GradleScopeCompileServices.configure(CompileServices.java:41)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.internal.service.ReflectionBasedServiceMethod.invoke(ReflectionBasedServiceMethod.java:35)
        at org.gradle.internal.service.DefaultServiceRegistry.applyConfigureMethod(DefaultServiceRegistry.java:195)
        ... 64 more
```
Any idea what I am doing wrong?

**Comment by fvclaus** [comment]
@FieteO , I can reproduce your error, if I ran Gradle with Java 17. You must use Java 11 for Gradle. This is my `gradle.properties`:
```
org.gradle.java.home=C://Users//fvclaus//.jabba//jdk//openjdk@1.17.0
```

File: docs/development/Setup.md (hunk 1 above)
Comment (fvclaus) [reply docs/development/Setup.md:72]: I added a section describing the Gradle setup, including the requirement to use JDK 11 for Gradle.

**Comment by FieteO** [comment]
> @FieteO , I can reproduce your error, if I run Gradle with Java 17. You must use Java 11 for Gradle. This is my `gradle.properties`:
> 
> ```
> org.gradle.java.home=C://Users//fvclaus//.jabba//jdk//openjdk@1.11.0-2
> ```

Hmm, my `JAVA_HOME` is set to java 11. The build goes further (but eventually fails) when I configure java 8 there.

File: docs/development/Setup.md [hunk 2]
```diff
@@ -56,14 +62,28 @@ finish before doing anything! Once finished, you should see the workspace below.
...
```
Comment (donat) [inline docs/development/Setup.md:71]:
```suggestion
The target platform selected during the installation can be changed. To do that, open one of the `tooling-eXXX.target` files and select `Set as Active Target Platform`
```

**Review by donat — APPROVED** [review]
The changes make sense, thanks for updating the docs.

**Pull request merged**